runtime/tsan.c

```diff
@@ -98,7 +98,99 @@
+
+   This mapping dictates which instrumentation calls are generated by the
+   compiler, but also how some functions of the runtime are instrumented: for
+   example, caml_modify is instrumented with a call to __tsan_write8, whereas
+   caml_initialize is purposefully not instrumented.
+
+   3.1. False negatives
+
+   There should be no false negatives, i.e., all data races (in the OCaml
+   sense) on visited code paths should be detected (modulo TSan limitations
+   such as the finite history of memory accesses).
+
+   3.2. False positives
+
+   The mapping that we use should not incur any false positives in pure OCaml
+   code, i.e., all data races reported between two accesses from OCaml are
+   true data races (in the OCaml sense of the term).
+
+   In mixed C-OCaml code, rare false positives may occur in the following cases:
+
+   - A value is initialized from C without using `caml_initialize` (allowed by
+     the FFI rules on the condition that the GC does not run between the
+     allocation and the end of initialization) and a conflicting access is made
+     from OCaml after publication to other threads. There is no data race due
+     to data dependencies, but TSan does not take data dependencies into
+     account.
+   - A field is accessed from C with `Field`, or more generally using a
+     `volatile value *` or a relaxed atomic access, and that field is modified
+     concurrently by OCaml code. Because caml_modify is instrumented as a plain
+     write for proper detection of OCaml races, this case is seen as a data
+     race.
+   - An OCaml heap value is accessed from C by two unordered, volatile
+     accesses, at least one of which is a write.
+
+   3.3. volatile accesses
+
+   We consider volatile accesses in C to behave like relaxed atomic accesses
+   from the point of view of data races (see [MMOC] comment in memory.c). It is
+   not trivial to explain that to TSan. Fortunately, both GCC and Clang have an
+   option to distinguish volatile writes in a custom way, by instrumenting them
+   with a call to a symbol that is left to us to implement. This option has been
+   introduced to support KCSan, the kernel concurrency sanitizer of the Linux
+   kernel. However, as KCSan is very different from TSan, volatile accesses are
+   instrumented in a way that makes it difficult to handle volatile writes the
+   way we would like. While gcc/clang replace atomic read/write operations with
+   TSan calls that update TSan's internal state and perform the actual memory
+   operation themselves, volatile read/write operations are merely decorated
+   with a TSan call, and then the actual operation is performed.
+
+   More details and examples can be found in PR #12681.
+
+  Our current make-do solution is that __tsan_volatile_readN performs a dummy
+  call to __tsan_atomic64_load, which is sufficient for correctness; and
+  __tsan_volatile_writeN simply calls __tsan_write8, i.e., volatile writes are
+  treated as plain writes. */
```

Comment:
I do not understand yet this part, could you clarify why having __tsan_atomic64_load vs. __tsan_write8 does what you want? (Is it related to the false positives mentioned in another PR? If so this might also be worth mentioning.)